xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,218 @@
+
+import com.google.api.expr.v1alpha1.Expr;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Descriptors.Descriptor;
+import io.envoyproxy.envoy.config.rbac.v2.Policy;
+import io.envoyproxy.envoy.config.rbac.v2.RBAC;
+import io.envoyproxy.envoy.config.rbac.v2.RBAC.Action;
+import io.grpc.xds.internal.rbac.engine.cel.Activation;
+import io.grpc.xds.internal.rbac.engine.cel.DefaultDispatcher;
+import io.grpc.xds.internal.rbac.engine.cel.DefaultInterpreter;
+import io.grpc.xds.internal.rbac.engine.cel.DescriptorMessageProvider;
+import io.grpc.xds.internal.rbac.engine.cel.Dispatcher;
+import io.grpc.xds.internal.rbac.engine.cel.IncompleteData;
+import io.grpc.xds.internal.rbac.engine.cel.Interpretable;
+import io.grpc.xds.internal.rbac.engine.cel.Interpreter;
+import io.grpc.xds.internal.rbac.engine.cel.InterpreterException;
+import io.grpc.xds.internal.rbac.engine.cel.RuntimeTypeProvider;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * CEL Evaluation Engine is part of the authorization framework in gRPC.
+ * CEL Evaluation Engine takes one or two Envoy RBAC policies as input
+ * and uses CEL library to evaluate the condition field 
+ * inside each RBAC policy based on the provided Envoy Attributes. 
+ * CEL Evaluation Engine will generate an authorization decision which
+ * could be ALLOW, DENY or UNKNOWN.
+ * 
+ * <p>Use as in:
+ * 
+ * <pre>
+ *  AuthorizationEngine engine = new AuthorizationEngine(
+ *      ImmutableList.of(rbacPolicies));
+ *  AuthorizationDecision result = engine.evaluate(new EvaluateArgs(call, headers));
+ * </pre>
+ */
+public class AuthorizationEngine<ReqT, RespT> {
+  private static final Logger log = Logger.getLogger(AuthorizationEngine.class.getName());
+  
+  /**
+   * RbacEngine is an inner class that holds RBAC action 
+   * and conditions of RBAC policy.
+   */
+  private static class RbacEngine {
+    @SuppressWarnings("UnusedVariable")
+    private final Action action;
+    private final ImmutableMap<String, Expr> conditions;
+
+    public RbacEngine(Action action, ImmutableMap<String, Expr> conditions) {
+      this.action = action;
+      this.conditions = conditions;
+    }
+  }
+
+  private final RbacEngine allowEngine;
+  private final RbacEngine denyEngine;
+
+  /**
+   * Creates a CEL-based Authorization Engine from a list of Envoy RBACs.
+   * The constructor can take either one or two Envoy RBACs.
+   * When it takes two RBACs, the order has to be a RBAC with DENY action
+   * followed by a RBAC with ALLOW action.
+   * @param rbacPolicies input Envoy RBAC list.
+   * @throws IllegalArgumentException if the user inputs an invalid RBAC list.
+   */
+  public AuthorizationEngine(ImmutableList<RBAC> rbacPolicies) throws IllegalArgumentException {
```

Comment:
Instead of accepting a list of size 1 or 2 why not accept 2 params: if both params present then 1st param is DENY RBAC and 2nd param ALLOW RBAC

Is there a reason for a list (such as future enhancements) which is not documented here?

Reply:
@ZhenLian Could you please take a look at this? Is it okay to change the list to two RBAC parameters?

Reply:
Yeah, that looks better. Actually, can I further suggest we have two construction functions? One only takes one RBAC, and it could be either DENY or ALLOW. the other takes two RBACs, and we will verify if the 1st param is DENY RBAC and 2nd param ALLOW RBAC.

Reply:
That's even better. In any case I don't see any need for a constructor that takes a list of policies.

Reply:
Fixed!